### ISSUE TITLE
Adding Error Types to Error States Zeroes FTA

### ISSUE THREAD

**Issue by brlarson** [osate/osate2#2111]
## Summary
When running FTA on a composite state, the resulting probability was zero.  So I created FTA_example.aadl (attached) which has the smallest difference between FTA that gets the correct result, and FTA that gets zero.

## Expected and Current Behavior
The only difference is error behavior state machine ErrorDetector has error types attached to its error states, while ErrorDetector_ok does not.

## Steps to Reproduce
Correct Behavior:  Instantiate iPCA_Safety.ok; select instance model for Run Fault Tree Analysis;
select Fault Contributor Trace and Show in graphical view; select state undetected_failure.
The error rate will be 1.3e-10.

Incorrect Behavior: Instantiate iPCA_Safety.i instead; all else same.
The error rate will be 0.0e+00.

The only difference is iPCA_Safety.i uses subcomponent iPCA_Error_Detector.i, which uses behavior ErrorDetector, which has error types on its error states,

while iPCA_Safety.ok uses subcomponent iPCA_Error_Detector.ok, which uses behavior ErrorDetector_ok, which does not have error types on its error states.

```aadl
package FTA_example
public
	with Base_Types;

flag renames data Base_Types::Boolean;

annex EMV2 {**
	error types
	  ThreadFault renames type ErrorLibrary::EarlyServiceTermination;  --thread fault halts thread
	  InternalError : type;  --an internal error was detected
	  DetectedFault : type extends InternalError;  --fault was detected (and announced)
	  POSTdetectedFault : type extends DetectedFault;  --power-on self-test detected a fault
	  RAMdetectedFault : type extends DetectedFault;  --RAM detected a fault
	  ROMdetectedFault : type extends DetectedFault;  --ROM detected a fault
	  CPUdetectedFault : type extends DetectedFault;  --CPU detected a fault
	  HWdetectedFault : type extends DetectedFault;  --HW self-test detected a fault
	  UndetectedFault : type extends InternalError;  --fault occurred, but not detected
	end types;
		
		error behavior ErrorDetector_ok
			states
				working: initial state;
				failed_detected: state;
				failed_undetected: state;
		end behavior;

	error behavior ErrorDetector
	  states 
	    working: initial state; 
	    failed_detected : state       {POSTdetectedFault, RAMdetectedFault, ROMdetectedFault, CPUdetectedFault, ThreadFault, HWdetectedFault};
	    failed_undetected : state {UndetectedFault};
	end behavior;
		
		--composite 3-state working-operational-fail
		error behavior CompositeFailUndetected
			states
				working: initial state;
				silent: state; --speaker failure so alarms can't be heard'
				undetected_failure: state; --system is impaired in some unimportant way, but fully functional
				failed: state; --non-functional
		end behavior;
**};

system iPCA_Safety --extends Safety::Safety  
end iPCA_Safety;

system implementation iPCA_Safety.i --extends Safety::Safety.i
	subcomponents
     error_detect : device  iPCA_Error_Detector.i;   
  annex EMV2 {**
    use behavior FTA_example::CompositeFailUndetected; 	
  	composite error behavior
  	states
  	  fd : [error_detect.failed_detected]->failed;
  	  fu : [error_detect.failed_undetected]->undetected_failure;
  	end composite;
  **};
end iPCA_Safety.i;

device iPCA_Error_Detector extends Error_Detector
	features 
		hw_error_detections : refined to feature group inverse of Operation_Safety_iPCA;
	properties
		Classifier_Substitution_Rule => Type_Extension;
end iPCA_Error_Detector;

device implementation iPCA_Error_Detector.i
  annex EMV2
  {** 
  use behavior FTA_example::ErrorDetector; 	
  error propagations
    hw_error_detections.post_fail: out propagation {FTA_example::UndetectedFault}; 
  end propagations;
  component error behavior
    events
      post_detected_failure : error event;
      post_not_detect_failure : error event;
	  transitions
	    working -[post_detected_failure]-> failed_detected;   
	    working -[post_not_detect_failure]-> failed_undetected;   
  end component;
  properties
	--power-on self-test detected failure
		EMV2::OccurrenceDistribution =>  
		  [ProbabilityValue => 1.2E-7; Distribution => Fixed;]
		  applies to post_detected_failure;
	--power-on self-test did not detect failure when it should have
		EMV2::OccurrenceDistribution =>  
		  [ProbabilityValue => 1.3E-10;	Distribution => Fixed;]
		  applies to post_not_detect_failure;
  	
  **};  --end of EMV2
end iPCA_Error_Detector.i; 


system implementation iPCA_Safety.ok 
	subcomponents
     error_detect : device  iPCA_Error_Detector.ok;  
  annex EMV2 {**
    use behavior FTA_example::CompositeFailUndetected; 	
  	composite error behavior
  	states
  	  fd : [error_detect.failed_detected]->failed;
  	  fu : [error_detect.failed_undetected]->undetected_failure;
  	end composite;
  **};
end iPCA_Safety.ok;


device implementation iPCA_Error_Detector.ok
  annex EMV2
  {** 
  use behavior FTA_Example::ErrorDetector_ok; 	
  error propagations
    hw_error_detections.post_fail: out propagation {FTA_Example::UndetectedFault}; 
  end propagations;
  component error behavior
    events
      post_detected_failure : error event;
      post_not_detect_failure : error event;
	  transitions
	    working -[post_detected_failure]-> failed_detected;   
	    working -[post_not_detect_failure]-> failed_undetected;   
  end component;
  properties
	--power-on self-test detected failure
		EMV2::OccurrenceDistribution =>
		  [ProbabilityValue => 1.2E-7; Distribution => Fixed;]
		  applies to  post_detected_failure;
	--power-on self-test did not detect failure when it should have
		EMV2::OccurrenceDistribution => 
		  [ProbabilityValue => 1.3E-10;	Distribution => Fixed;]
		  applies to  post_not_detect_failure;
  	
  **};  --end of EMV2
end iPCA_Error_Detector.ok; 

feature group Operation_Safety
  --signals between the operation and safety subsystems
end Operation_Safety;


	feature group Operation_Safety_iPCA  extends Operation_Safety
		--signals between the operation and safety subsystems
		features
			Halt: in event port;
			pump_at_KVO_rate: in event port;
			priming_failure: out event port;
			drug_not_in_library: out event port;
			hard_limit_violated: out event port;
			over_max_drug_per_hour: out event port;
			security_fault: out event port;
			cp_reset_alarm: out event port;
			post_fail: in event port;
			post_done: in event port;
			hw_detected_failure: in event port;
	end Operation_Safety_iPCA;

device Error_Detector
	features
	  illuminate: out data port flag; --turn-on led
    hardware_fault : out event data port;   --tell logger hardware fault detected
    halt : out event port;  --halt medical device operation on hardware command
	  hw_error_detections:  feature group inverse of Operation_Safety;     --monitor thread heartbeat
end Error_Detector;

end FTA_example;
```

## Environment

* **OSATE Version**: 2.6.1
* **Operating System**: macOS 10.14.6

**Comment by reteprelief** [comment]
Hi Brian,
in the case where a probability value shows you have a state without type and a transition from an event (without type) to that state. As a result the fault tree traces back to the event and shows its probability.

In the case where the error state is defined to carry error types the fault tree/trace shows the error state (in the case of Failed state a separate state for each error type). It then tries to trace back to the error event. It does not find a transition from an error event that results in the error state with the particular error type, thus, does not add the error event and does not use its probability. 
At the moment in the case of unhandled_failure the subevent representing the error state for the detector does not display correctly - it does not show the correct object (Detector) and it skips showing the error type (UndetectedFault).
I will fix that issue.

On your end you need to modify your model to have error events with types - otherwise we do not know which type goes along with the error state, i.e., we do not find the error event when tracing back from the state to the triggering event specified in the transition.

**Comment by reteprelief** [comment]
The problems raised by this issue will be addressed in issues #2123, #2124, and #2125

**Comment by brlarson** [comment]
Thank you Peter.  I will add error types to all error events.
